### PR TITLE
Refactor E2E test code to avoid repetition

### LINKED ISSUE #755: Refactor E2E test setup code to reduce repetition
In a lot of places, we have various tx/deployments/other setup boilerplate at the start of our E2E tests. These should be moved into proper setup functions.

Related: https://github.com/cowprotocol/services/pull/746#discussion_r1017063136.

### PR DESCRIPTION
**Pull request by Razz4780**
Fixes #755.

This PR introduces some extra helpers to the e2e code to reduce repetition and make writing new tests more convenient. New `OnchainComponents` provides methods for batch generating user and solver accounts. `Services` struct provides methods for starting the offchain services and interacting with them.

I also updated `README.md` part regarding hardhat setup, it didn't include the gas configuration.

### Test Plan

Existing e2e tests still pass.

### CODE REVIEW

**Comment by Razz4780** [comment]
When doing the refactor, I noticed that e2e run sequentially, because they use the database and the hardhat node as shared resources. We could use environment variables to optionally support multiple databases and nodes, e.g.
```
DB_NAMES=db_1,db_2,db_3
NODE_PORTS=8545,8546,8547
```

File: crates/e2e/tests/e2e/helpers/onchain_components.rs [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+
+        Self {
+            web3,
+            contracts,
+            accounts: Default::default(),
+        }
+    }
+
+    /// Generate next `N` accounts with the given initial balance.
+    pub async fn make_accounts<const N: usize>(&mut self, with_wei: U256) -> [TestAccount; N] {
```
Comment (MartinquaXD) [inline crates/e2e/tests/e2e/helpers/onchain_components.rs:193]: Nice use of const generics!

File: crates/e2e/tests/e2e/helpers/services.rs [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+/// Exposes various utility methods for tests.
+pub struct Services<'a> {
+    contracts: &'a Contracts,
+    http: Client,
+}
+
+impl<'a> Services<'a> {
+    /// Create a new instance of this struct and clear the database.
+    pub async fn new(contracts: &'a Contracts) -> Services<'a> {
+        Self::clear_database().await;
```
Comment (MartinquaXD) [inline crates/e2e/tests/e2e/helpers/services.rs:36]: Not sure if clearing a DB should be part of this constructor. I get that it should happen for every test and this would be an easy way to that under the hood but these 2 things seem unrelated.
I think it would be different if the `Services` struct would create a fresh in-memory DB instance and made every component use that instead. Not sure how easy that would be, though.

File: crates/e2e/tests/e2e/helpers/services.rs [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+
+        match status {
+            StatusCode::OK => Ok(serde_json::from_str(&body).unwrap()),
+            code => Err((code, body)),
+        }
+    }
+
+    pub async fn solvable_orders(&self) -> usize {
+        self.get_auction().await.auction.orders.len()
+    }
```
Comment (MartinquaXD) [inline crates/e2e/tests/e2e/helpers/services.rs:220]: I think it would be a bit nicer to return the orders themselves instead of just the count. My reasoning is that manually calling `.len()` is not a lot of overhead and wouldn't require another function in case you want to assert something about the contents of the orders.

**Review by MartinquaXD — COMMENTED** [review]
Really nice PR! 👍 
Didn't review everything yet but the idea is great. I always dreaded writing these tests and this would make it sooo much nicer.

File: crates/e2e/tests/e2e/helpers/mod.rs [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+mod services;
+
+use {
+    anyhow::{anyhow, Result},
+    std::{future::Future, io::Write, time::Duration},
+    tempfile::TempPath,
+};
+pub use {deploy::*, onchain_components::*, services::*};
+
+/// Initialize tracing and assure that panic hook is set.
```
Comment (MartinquaXD) [inline crates/e2e/tests/e2e/helpers/mod.rs:13]: Currently doesn't set the panic hook. Is that intended?

File: crates/e2e/tests/e2e/colocation_univ2.rs [hunk 5]
```diff
@@ -25,80 +20,46 @@ async fn local_node_test() {
...
-            token_amount: to_wei(1000),
-            weth_amount: to_wei(1000),
-        },
-    )
-    .await;
-    let token = token.contract;
+    let mut onchain = OnchainComponents::deploy(web3).await;
+
+    let [solver] = onchain.make_solvers(to_wei(10)).await;
+    let [trader] = onchain.make_accounts(to_wei(10)).await;
```
Comment (nlordell) [inline crates/e2e/tests/e2e/colocation_univ2.rs:29]: Woah! I didn't know this hit stable already. Nice!

File: crates/e2e/tests/e2e/helpers/services.rs (hunk 2 above)
Comment (Razz4780) [reply crates/e2e/tests/e2e/helpers/services.rs:36]: Yeah, maybe doing this here is not very obvious. Maybe we could move clearing the database to the test wrapper function? That function currently locks and cleans up the hardhat node.
Creating a new DB instance for each test is a bit problematic - we can either use the existing database as a template and clone it inside Postgres (this fails if there are any existing connections to the template database) or create a new database and run migrations (not sure how our migrations work, as they use flyway)

File: crates/e2e/tests/e2e/helpers/services.rs (hunk 3 above)
Comment (Razz4780) [reply crates/e2e/tests/e2e/helpers/services.rs:220]: `solvable_orders` was a standalone function returning the size of the current auction. There is another method `get_auction`, which returns the current auction itself. I thought about removing this entirely and using `services.get_auction().await.auction.orders.len()` everywhere, but in many places we're interested only with the count of the orders. So I left it this way :p

File: crates/e2e/tests/e2e/helpers/mod.rs (hunk 4 above)
Comment (Razz4780) [reply crates/e2e/tests/e2e/helpers/mod.rs:13]: Thanks, fixed

File: crates/e2e/tests/e2e/helpers/deploy.rs [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    model::DomainSeparator,
+    shared::ethrpc::Web3,
+};
+
+pub struct Contracts {
+    pub balancer_vault: BalancerV2Vault,
+    pub gp_settlement: GPv2Settlement,
+    pub gp_authenticator: GPv2AllowListAuthentication,
+    pub uniswap_factory: UniswapV2Factory,
+    pub uniswap_router: UniswapV2Router02,
```
[the remark this answers is not in the document]
Comment (Razz4780) [reply crates/e2e/tests/e2e/helpers/deploy.rs]: Sure, done

File: crates/e2e/tests/e2e/helpers/onchain_components.rs [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+}
+
+impl Iterator for AccountGenerator {
+    type Item = TestAccount;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            self.id = self.id.checked_add(1)?;
+
+            let mut buffer = [0; 32];
```
[the remark this answers is not in the document]
Comment (Razz4780) [reply crates/e2e/tests/e2e/helpers/onchain_components.rs]: Thanks, done

File: crates/e2e/tests/e2e/helpers/onchain_components.rs [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+    let mut bytes = vec![0u8; 65];
+    bytes[0..32].copy_from_slice(signature.r.as_bytes());
+    bytes[32..64].copy_from_slice(signature.s.as_bytes());
+    bytes[64] = signature.v as _;
+
+    bytes
+}
+
+pub fn to_wei(base: u32) -> U256 {
+    U256::from(base) * U256::from(10).pow(18.into())
```
[the remark this answers is not in the document]
Comment (Razz4780) [reply crates/e2e/tests/e2e/helpers/onchain_components.rs]: Thanks, done

File: crates/e2e/tests/e2e/helpers/onchain_components.rs [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+struct AccountGenerator {
+    id: usize,
+}
+
+impl Iterator for AccountGenerator {
+    type Item = TestAccount;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            self.id = self.id.checked_add(1)?;
```
[the remark this answers is not in the document]
Comment (Razz4780) [reply crates/e2e/tests/e2e/helpers/onchain_components.rs:138]: Kinda disagree, it's a pretty natural thing for an `Iterator` to run out of items. Signatures of `OnchainComponents` methods guarantee to return the required count of items anyway

[the remark this answers is not in the document]
Comment (MartinquaXD) [reply crates/e2e/tests/e2e/helpers/onchain_components.rs:138]: Ah sorry. You are totally right. For some reason I thought `self.id.checked_add(1)?;` would cause the `id` to wrap over and restart the loop. 🤦‍♂️

File: crates/e2e/tests/e2e/helpers/deploy.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (MartinquaXD) [reply crates/e2e/tests/e2e/helpers/deploy.rs]: Nice only 1 thing left: `deploy_tokens_with_weth_uni_pools -> deploy_tokens_with_weth_uni_v2_pools`

File: crates/e2e/tests/e2e/helpers/deploy.rs [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+        let gp_settlement = deploy!(GPv2Settlement(
+            gp_authenticator.address(),
+            balancer_vault.address(),
+        ));
+
+        gp_authenticator
+            .add_solver(admin)
+            .send()
+            .await
+            .expect("failed to allow list account 0");
```
Comment (vkgnosis) [inline crates/e2e/tests/e2e/helpers/deploy.rs:82]: Is not needed because we also have the make_solver function.

**Review by vkgnosis — APPROVED** [review]

**Review by nlordell — APPROVED** [review]
Great improvements! I couldn't review the entire PR in great detail (it was a bit long), but what I chose to review looks good.

File: crates/e2e/tests/e2e/helpers/deploy.rs [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+
+impl Contracts {
+    pub async fn deploy(web3: &Web3) -> Self {
+        let network_id = web3.net().version().await.expect("get network ID failed");
+        tracing::info!("connected to test network {}", network_id);
+
+        let accounts: Vec<Address> = web3.eth().accounts().await.expect("get accounts failed");
+        let admin = accounts[0];
+
+        macro_rules! deploy {
```
[the remark this answers is not in the document]
Comment (vkgnosis) [reply crates/e2e/tests/e2e/helpers/deploy.rs:37]: I don't like macros either so I'm fine with removing them. Usually you can write generic functions that only have small overhead over macros. Also fine with keeping them because this PR isn't the one that introduced them.

File: crates/e2e/tests/e2e/helpers/services.rs (hunk 2 above)
Comment (Razz4780) [reply crates/e2e/tests/e2e/helpers/services.rs]: Done ^^

File: crates/e2e/tests/e2e/helpers/deploy.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Razz4780) [reply crates/e2e/tests/e2e/helpers/deploy.rs]: Doone

File: crates/e2e/tests/e2e/helpers/deploy.rs (hunk 10 above)
Comment (Razz4780) [reply crates/e2e/tests/e2e/helpers/deploy.rs]: Removed, thanks

File: crates/e2e/tests/e2e/helpers/mod.rs [hunk 12]
```diff
@@ -0,0 +1,53 @@
+mod deploy;
```
[the remark this answers is not in the document]
Comment (Razz4780) [reply crates/e2e/tests/e2e/helpers/mod.rs]: Done, renamed to `setup` ^^

**Review by MartinquaXD — APPROVED** [review]

File: crates/e2e/tests/e2e/helpers/deploy.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (Razz4780) [reply crates/e2e/tests/e2e/helpers/deploy.rs]: This macro was already there. I can remove it (along with the `tx...` macros), but maybe in another PR?

File: crates/e2e/tests/e2e/helpers/services.rs [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+        let status = placement.status();
+        let body = placement.text().await.unwrap();
+
+        match status {
+            StatusCode::CREATED => Ok(serde_json::from_str(&body).unwrap()),
+            code => Err((code, body)),
+        }
+    }
+
+    /// Submit an [model::quote::OrderQuote].
```
[the remark this answers is not in the document]
Comment (Razz4780) [reply crates/e2e/tests/e2e/helpers/services.rs]: Done

File: crates/e2e/tests/e2e/helpers/services.rs [hunk 14]
```diff
@@ -0,0 +1,244 @@
...
+
+        let status = response.status();
+        let body = response.text().await.unwrap();
+
+        assert_eq!(status, StatusCode::OK, "{body}");
+
+        serde_json::from_str(&body).unwrap()
+    }
+
+    /// Create an [Order].
```
[the remark this answers is not in the document]
Comment (Razz4780) [reply crates/e2e/tests/e2e/helpers/services.rs]: Done
